File: foxogram-common/src/main/java/su/foxogram/dtos/api/response/ChannelDTO.java
package su.foxogram.dtos.api.response;

import com.fasterxml.jackson.annotation.JsonInclude;
import io.swagger.v3.oas.annotations.media.Schema;
import lombok.Getter;
import lombok.Setter;
import su.foxogram.constants.ChannelsConstants;
import su.foxogram.models.Channel;
import su.foxogram.models.Message;

@Getter
@Setter
@Schema(name = "Channel")
public class ChannelDTO {

	private long id;

	private String displayName;

	private String name;

	private String icon;

	private int type;

	private boolean isPublic;

	private long flags;

	private int memberCount;

	private UserDTO owner;

	private long createdAt;

	@JsonInclude(JsonInclude.Include.NON_NULL)
	private MessageDTO lastMessage;

	public ChannelDTO(Channel channel, Message lastMessage) {
		this.id = channel.getId();
		this.displayName = channel.getDisplayName();
		this.name = channel.getName();
		this.icon = channel.getIcon();
		this.type = channel.getType();
		this.flags = channel.getFlags();
		this.isPublic = channel.hasFlag(ChannelsConstants.Flags.PUBLIC);
		if (channel.getMembers() != null) {
			this.memberCount = channel.getMembers().size();
		}
		if (lastMessage != null) {
			this.lastMessage = new MessageDTO(lastMessage, null);
		}
		this.owner = new UserDTO(channel.getOwner(), null, false, false);
		this.createdAt = channel.getCreatedAt();
	}
}

File: src/main/java/su/foxogram/dto/api/response/ChannelDTO.kt
package su.foxogram.dto.api.response

import com.fasterxml.jackson.annotation.JsonInclude
import io.swagger.v3.oas.annotations.media.Schema
import lombok.Getter
import lombok.Setter
import su.foxogram.model.Channel
import su.foxogram.model.Message

@Getter
@Setter
@Schema(name = "Channel")
data class ChannelDTO(val channel: Channel, val message: Message?) {
    val id: Long = channel.getId()

    val displayName: String? = channel.displayName

    val name: String? = channel.name

    var icon: AttachmentDTO? = if (channel.icon != null) AttachmentDTO(channel.icon) else null

    val type: Int = channel.type

    val flags: Long = channel.flags

    var memberCount = 0

    val owner: UserDTO? = UserDTO(
        channel.owner, null, null,
        includeEmail = false,
        includeChannels = false,
        includeContacts = false
    )

    val createdAt: Long = channel.createdAt

    @JsonInclude(JsonInclude.Include.NON_NULL)
    var lastMessage: MessageDTO? = if (message != null) MessageDTO(message, false) else null
}
